### PR TITLE
Add support for Voice Settings

### PR DESCRIPTION
**Pull request by Mr-Pine**
This PR adds support for Voice Settings by 

- Implementing the `GET_VOICE_SETTINGS` and `SET_VOICE_SETTINGS` commands
- Implementing the `VOICE_SETTINGS_CHANGED` Event
- Implementing the authorization flow enabling different scopes by adding the `AUTHORIZE` command
- Renamed `getOAuthToken` to `authenticate` to represent the actual command name
- Changed example project to contain getting and setting voice settings
- Added a gitignored `secrets.kt` containing client id and secret

### CODE REVIEW

File: src/example/kotlin/dev/cbyrne/example/Main.kt [hunk 1]
```diff
@@ -66,5 +169,47 @@ suspend fun main() {
...
+    constructor(response: OAuthResponse) : this(
+        accessToken = response.accessToken,
+        tokenType = response.tokenType,
+        refreshToken = response.refreshToken,
+        scope = response.scope,
+        expiresOn = Clock.System.now() + response.expiresIn.seconds
+    )
+}
+
+fun saveOAuthResponse(response: OAuthResponse) {
```
Comment (caoimhebyrne) [inline src/example/kotlin/dev/cbyrne/example/Main.kt]: This code is not platform agnostic.

File: src/main/kotlin/dev/cbyrne/kdiscordipc/core/packet/outbound/impl/VoiceSettingsPacket.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class GetVoiceSettingsPacket(
+    override val opcode: Int = 0x01,
+    override val cmd: String = "GET_VOICE_SETTINGS",
+    override val args: Arguments? = null,
+    override var nonce: String = "0"
+): OutboundPacket(){
+}
```
Comment (caoimhebyrne) [inline src/main/kotlin/dev/cbyrne/kdiscordipc/core/packet/outbound/impl/VoiceSettingsPacket.kt]: ``{}`` is not required here?

File: src/main/kotlin/dev/cbyrne/kdiscordipc/core/packet/outbound/impl/VoiceSettingsPacket.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+@Serializable
+data class SetVoiceSettingsPacket(
+    override val opcode: Int = 0x01,
+    override val cmd: String = "SET_VOICE_SETTINGS",
+    override val args: VoiceSettingArguments,
+    override var nonce: String = "0"
+): OutboundPacket(){
+
```
Comment (caoimhebyrne) [inline src/main/kotlin/dev/cbyrne/kdiscordipc/core/packet/outbound/impl/VoiceSettingsPacket.kt]: Remove whitespace

File: src/main/kotlin/dev/cbyrne/kdiscordipc/core/util/temporaryDirectory.kt [hunk 4]
```diff
@@ -1,4 +1,4 @@
 package dev.cbyrne.kdiscordipc.core.util
 
 val temporaryDirectory =
-    System.getenv("XDG_RUNTIME_DIR") ?: System.getenv("TMPDIR") ?: System.getenv("TMP") ?: System.getenv("TMPDIR")
```
Comment (caoimhebyrne) [inline src/main/kotlin/dev/cbyrne/kdiscordipc/core/util/temporaryDirectory.kt:4]: Why was this removed?

File: src/main/kotlin/dev/cbyrne/kdiscordipc/data/voiceSettings/VoiceSettings.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    override val deaf: Boolean
+): EventData(), VoiceSettingsInterface
+
+@Serializable
+data class InputOutput(
+    @SerialName("available_devices") val availableDevices: Array<Device>? = null,
+    @SerialName("device_id") val deviceId: String? = null,
+    val volume: Float? = null
+){
+
```
Comment (caoimhebyrne) [inline src/main/kotlin/dev/cbyrne/kdiscordipc/data/voiceSettings/VoiceSettings.kt]: Remove this whitespace

File: src/main/kotlin/dev/cbyrne/kdiscordipc/data/voiceSettings/VoiceSettings.kt [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    @Serializable
+    data class Device(
+        val id: String,
+        val name: String
+    )
+}
+
```
Comment (caoimhebyrne) [inline src/main/kotlin/dev/cbyrne/kdiscordipc/data/voiceSettings/VoiceSettings.kt]: Remove this whitespace

File: src/main/kotlin/dev/cbyrne/kdiscordipc/data/voiceSettings/VoiceSettings.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+@Serializable
+data class Mode(
+    val type: String,
+    @SerialName("auto_threshold") val autoThreshold: Boolean,
+    val threshold: Float,
+    val shortcut: Array<ShortCutKeyCombo>,
+    val delay: Float
+) {
+
```
Comment (caoimhebyrne) [inline src/main/kotlin/dev/cbyrne/kdiscordipc/data/voiceSettings/VoiceSettings.kt]: Remove this whitespace

File: src/example/kotlin/dev/cbyrne/example/Main.kt [hunk 8]
```diff
@@ -43,9 +57,98 @@ suspend fun main() {
...
+
+            val oauthResponse = json.decodeFromString<OAuthResponse>(response.body!!.string())
+            logger.info(oauthResponse)
+
+            accessToken = oauthResponse.accessToken
+
+            saveOAuthResponse(oauthResponse)
+        }
+
+        // Authenticate with the client and get an oauth token for the currently logged-in user
```
Comment (caoimhebyrne) [inline src/example/kotlin/dev/cbyrne/example/Main.kt]: I'm not a fan of the HTTP request code used here. Can we use ktor and make it look a little nicer please?

File: src/example/kotlin/dev/cbyrne/example/Main.kt [hunk 9]
```diff
@@ -43,9 +57,98 @@ suspend fun main() {
...
+        val oauthToken = ipc.applicationManager.authenticate(accessToken)
         logger.info("Received oauth token from Discord! Expires on: ${oauthToken.expires}")
+
+        //Prints the current voice settings
+        val voiceSettings = ipc.voiceSettingsManager.getVoiceSettings()
+        logger.info("Voice Settings: $voiceSettings")
+
+        //Mutes the user
```
Comment (caoimhebyrne) [inline src/example/kotlin/dev/cbyrne/example/Main.kt]: Put spaces before the content of the comment. (i.e. ``//Test``should be``// Test``)

**Review by caoimhebyrne — CHANGES_REQUESTED** [review]
Just a few minor changes required.

File: src/main/kotlin/dev/cbyrne/kdiscordipc/core/util/temporaryDirectory.kt (hunk 4 above)
Comment (Mr-Pine) [reply src/main/kotlin/dev/cbyrne/kdiscordipc/core/util/temporaryDirectory.kt:4]: While looking through the code I saw this and thought it was not necessary to check "TMPDIR" twice. If it isn't a mistake, why is it necessary here?

**Comment by Mr-Pine** [comment]
Adressed all issues except the temp directory where I don't understand why the old code is better (added a more detailed code as a comment to the change request). Thanks for the quick feedback!

File: src/main/kotlin/dev/cbyrne/kdiscordipc/core/util/temporaryDirectory.kt (hunk 4 above)
Comment (caoimhebyrne) [reply src/main/kotlin/dev/cbyrne/kdiscordipc/core/util/temporaryDirectory.kt:4]: Oh oops, an oversight on my part.

**Review by caoimhebyrne — APPROVED** [review]
LGTM!
